walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,116 @@
+    HashSet<Event> eventsSet = new HashSet<>(events);
+
+    ArrayList<TimeRange> mandatoryTimes = eventsSet
+      .stream()
+      .filter(event -> this.checkAttendees(event, request, false) == true)
+      .map(event -> event.getWhen())
+      .collect(Collectors.toCollection(ArrayList::new));
+
+    ArrayList<TimeRange> optionalTimes = eventsSet
+      .stream()
+      .filter(event -> this.checkAttendees(event, request, true) == true)
+      .map(event -> event.getWhen())
+      .collect(Collectors.toCollection(ArrayList::new));
+
+    System.out.println(mandatoryTimes);
+    System.out.println(optionalTimes);
+
+    //sort time ranges by start
+    Collections.sort(mandatoryTimes, TimeRange.ORDER_BY_START);
+    Collections.sort(optionalTimes, TimeRange.ORDER_BY_START);
+
+    ArrayList<TimeRange> mandatoryFreeTimes = findFreeTimes(mandatoryTimes, request);
+    ArrayList<TimeRange> optionalFreeTimes = findFreeTimes(optionalTimes, request);
+
+    if (optionalFreeTimes.isEmpty()) {
+      return mandatoryFreeTimes;
+    }
+    
+    return optionalFreeTimes;
+  }
+
+  private static boolean checkAttendees(Event event, MeetingRequest request, boolean checkOptional) {
+    HashSet<String> attendees = new HashSet<>(event.getAttendees());
+    HashSet<String> requested = new HashSet<>(request.getAttendees());
+    HashSet<String> optional = new HashSet<>(request.getOptionalAttendees());
+
+    for (String person: requested) {
+      if (attendees.contains(person)) {
+        return true;
+      }
+    }
+    
+    if (checkOptional) {
+      for (String person: optional) {
+        if (attendees.contains(person))
+          return true;
+      }
+    }
+
+    return false;
+  }
+
+  private static ArrayList<TimeRange> findFreeTimes(ArrayList<TimeRange> times, MeetingRequest request) {
+    //get free times
+    ArrayList<TimeRange> freeTimes = new ArrayList();
+    
+    if (times.isEmpty() && (request.getDuration() <= totalMinutes)) {
+      freeTimes.add(TimeRange.WHOLE_DAY);
+      return freeTimes;
+    }
+
+    if (times.isEmpty()) {
+      return freeTimes;
+    }
+
+    TimeRange intersection = times.get(0);
+
+    for (TimeRange range: times) {
+      if (intersection.contains(range)) {
+        continue;
+      } else {
+        if (intersection.overlaps(range)) {
+          intersection = new TimeRange(intersection.start(), range.end() - intersection.start());
+        } else {
+          //found free time
+          TimeRange free = new TimeRange(intersection.end(), range.start() - intersection.end());
+          intersection = new TimeRange(range.start(), range.end() - range.start());
+          if (free.duration() >= request.getDuration()) {
+            freeTimes.add(free);
+          }
+        }
+      }
+    }
+
+    //check for beginning and end
+    if (TimeRange.START_OF_DAY < times.get(0).start()) {
+      freeTimes.add(new TimeRange(0, times.get(0).start()));
```

Comment:
nitpick: what if `request.getDuration()` is more than the time time between `TimeRange.START_OF_DAY` and `times.get(0).start()`?

Reply:
Thanks for pointing it out! Forgot to think about that case.